### PR TITLE
Adding more token encoding types

### PR DESCRIPTION
**Pull request by snarayan21**
Depending on the vocab size, users can encode their token IDs using various int formats. Previously, we only allowed for int64, which covers an absurdly high vocab size. Enabling decoding tokens in uint32 or  uint16, for example, would let people save space on their datasets since the max vocab sizes supported would be ~4 million with uint32, or ~65k with uint16. This has been added to both the text and finetuning datasets.

This PR also lets users specify their MDS dataset columns using `ndarray` types to enable automatically encoding/decoding samples. This was already present for finetuning dataset, so the functionality has been added for the generic text dataset. Accordingly, I've changed the default value in our MDS conversion scripts to use `ndarray:uint32` instead of `bytes` and made relevant changes to get this working.

Added unit tests checking that this works for text and finetuning datasets and that an error is thrown for uncompatible encoding types.

Moved a util function that was applicable to both the text and finetuning dataloaders to a common location to import. This had been written twice to perform the same functionality.

Ran the following scripts successfully: `convert_dataset_hf.py`, `convert_dataset_json.py`, `convert_finetuning_dataset.py`, `convert_text_to_mds.py`. Updated data prep readme to have instructions for `convert_text_to_mds.py`.

Using the shakespeare text file [here](https://ocw.mit.edu/ans7870/6/6.006/s08/lecturenotes/files/t8.shakespeare.txt), models trained with/without this branch have deterministic loss curves. One set of runs was with global batch size 32, the other with global batch size 256. See wandb project [here](https://wandb.ai/mosaic-ml/foundry-data-prep-runs?nw=nwusersaaketh).

Foundry regression tests are partially borked right now because of a small bug that's getting addressed in the release branch, but the tests that did run all succeeded. See [here](https://databricks.slack.com/archives/C05T1A4UMT8/p1717694690095049).

### CODE REVIEW

File: llmfoundry/data/text_data.py [hunk 1]
```diff
@@ -106,6 +106,7 @@ def __init__(
         self,
         tokenizer: PreTrainedTokenizerBase,
         max_seq_len: int,
+        token_encoding_type: str = 'int32',
```
Comment (codestar12) [inline llmfoundry/data/text_data.py]: can we make default 'int64' so it doesn't break compatibility with old code

**Review by codestar12 — CHANGES_REQUESTED** [review]
Looks good other then default value

File: llmfoundry/data/text_data.py (hunk 1 above)
Comment (XiaohanZhangCMU) [reply llmfoundry/data/text_data.py]: I favor 64 too, at least for the current and coming release.

**Review by XiaohanZhangCMU — APPROVED** [review]
LGTM. Suggest keeping the previous default for now.

File: llmfoundry/data/text_data.py (hunk 1 above)
Comment (snarayan21) [reply llmfoundry/data/text_data.py]: resolved

**Review by codestar12 — APPROVED** [review]
looks good

File: llmfoundry/data/text_data.py [hunk 2]
```diff
@@ -137,13 +140,30 @@ def __init__(
...
             )
 
-        if local is not None and (remote is None or (local == remote)):
-            if os.path.isdir(local):
-                contents = set(os.listdir(local))
-                if split not in contents:
-                    raise ValueError(
-                        f'local directory {local} does not contain split {split}',
-                    )
+        if token_encoding_type not in [
```
Comment (XiaohanZhangCMU) [inline llmfoundry/data/text_data.py:143]: also "int"

File: llmfoundry/data/data.py [hunk 3]
```diff
@@ -153,3 +153,17 @@ def __iter__(self) -> Iterable[Dict[str, bytes]]:
                     # convert to bytes to store in MDS binary format
                     'tokens': np.asarray(concat_sample).tobytes(),
                 }
+
+
+def stream_remote_local_validate(
```
[the remark this answers is not in the document]
Comment (snarayan21) [reply llmfoundry/data/data.py:158]: done

File: llmfoundry/data/finetuning/tasks.py [hunk 4]
```diff
@@ -606,11 +596,26 @@ def __init__(
                 f'StreamingFinetuningDataset() got an unexpected keyword argument: {kwargs}',
             )
 
+        if token_encoding_type not in [
```
[the remark this answers is not in the document]
Comment (snarayan21) [reply llmfoundry/data/finetuning/tasks.py:599]: done

File: llmfoundry/data/text_data.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (snarayan21) [reply llmfoundry/data/text_data.py:143]: done

File: llmfoundry/data/text_data.py [hunk 5]
```diff
@@ -41,6 +41,8 @@ class StreamingTextDataset(StreamingDataset):
         tokenizer (Tokenizer): HuggingFace tokenizer to
             tokenize samples.
         max_seq_len (int): The max sequence length of each sample.
+        token_encoding_type (str): The encoding type of the tokenized samples. Can be one of
```
[the remark this answers is not in the document]
Comment (snarayan21) [reply llmfoundry/data/text_data.py:44]: done

[the remark this answers is not in the document]
Comment (snarayan21) [reply llmfoundry/data/text_data.py:44]: So I'm not sure it's possible to include a variable in a docstring...I can list all of the types out, or I can just specify the default. Which would you prefer

File: tests/data/test_data_encodings.py [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+    dataset_local_path = str(tmp_path)
+    encoding_dtype = getattr(np, token_encoding_type)
+
+    if use_bytes:
+        columns = {
+            'tokens': 'bytes',
+        }
+    else:
+        columns = {
+            'tokens': 'ndarray:' + token_encoding_type,
```
[the remark this answers is not in the document]
Comment (snarayan21) [reply tests/data/test_data_encodings.py:37]: yep!

File: llmfoundry/data/text_data.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (dakinggg) [reply llmfoundry/data/text_data.py:44]: oh sorry, the "same" here just referred to adding the default. No need to try to put a variable in the docstring (although I'm pretty sure this actually possible).

[the remark this answers is not in the document]
Comment (snarayan21) [reply llmfoundry/data/text_data.py]: You can but it can get a bit ugly, it seems https://stackoverflow.com/questions/36091491/python-how-does-one-include-a-variable-in-the-doc-string

[the remark this answers is not in the document]
Comment (snarayan21) [reply llmfoundry/data/text_data.py]: But added the default

File: llmfoundry/data/finetuning/tasks.py [hunk 7]
```diff
@@ -494,26 +498,13 @@ def is_valid_ift_example(
...
-                )
-
-
 class StreamingFinetuningDataset(StreamingDataset):
     """Finetuning dataset with flexible tokenization using StreamingDataset.
 
     Args:
         tokenizer (Tokenizer): The name of the HuggingFace tokenizer to use to
             tokenize samples.
+        token_encoding_type (str): The encoding type of the tokenized samples. Defaults to 'int64'.
```
Comment (dakinggg) [inline llmfoundry/data/finetuning/tasks.py:507]: mention that this is only applied if saved direclty as bytes. if saved as numpy ,correct type will be inferred.

File: llmfoundry/data/text_data.py [hunk 8]
```diff
@@ -41,6 +44,7 @@ class StreamingTextDataset(StreamingDataset):
         tokenizer (Tokenizer): HuggingFace tokenizer to
             tokenize samples.
         max_seq_len (int): The max sequence length of each sample.
+        token_encoding_type (str): The encoding type of the tokenized samples. Defaults to 'int64'.
```
Comment (dakinggg) [inline llmfoundry/data/text_data.py:47]: same

File: llmfoundry/data/text_data.py [hunk 9]
```diff
@@ -197,10 +210,18 @@ def _read_binary_tokenized_sample(
...
     ) -> torch.Tensor:
-        return torch.from_numpy(
-            np.frombuffer(sample['tokens'],
-                          dtype=np.int64)[:self.max_seq_len].copy(),
-        )
+        # Modeling code still expects int64 tensors.
+        if isinstance(sample['tokens'], np.ndarray):
+            return torch.from_numpy(
+                sample['tokens'][:self.max_seq_len].copy(),
+            ).to(torch.int64)
```
Comment (dakinggg) [inline llmfoundry/data/text_data.py:217]: jw, was this just a typing thing? or did it actually error?

File: scripts/data_prep/convert_dataset_json.py [hunk 10]
```diff
@@ -161,7 +163,11 @@ def generate_samples(
             if truncate_num_samples is not None and n_samples == truncate_num_samples:
                 return
             n_samples += 1
-            yield {k: v[idx] for k, v in batch.items()}
+            yield {
+                k:
+                v[idx].numpy() if isinstance(v[idx], torch.Tensor) else v[idx]
```
Comment (dakinggg) [inline scripts/data_prep/convert_dataset_json.py:168]: where do the tensors come from? in fact I don't think this function is even used anywhere?

File: tests/a_scripts/data_prep/test_convert_text_to_mds.py [hunk 11]
```diff
@@ -194,7 +193,8 @@ def call_convert_text_to_mds() -> None:
     n_tokens = 0
     for i in range(dataset.num_samples):
         sample = dataset[i]
-        tokens = np.frombuffer(sample['tokens'], dtype=int)
+        # TODO: delete the below
```
Comment (dakinggg) [inline tests/a_scripts/data_prep/test_convert_text_to_mds.py:196]: whats the todo for?

**Review by dakinggg — COMMENTED** [review]
lgtm, I think it would be good to manual test at least one of the scripts? maybe do a training run using the new version of convert text to mds and make sure it still trains? and run the regression test suite to make sure existing datasets didn't get broken.

File: llmfoundry/data/text_data.py (hunk 9 above)
Comment (snarayan21) [reply llmfoundry/data/text_data.py:217]: it actually errors because torch's CrossEntropyLoss expects labels to have `int64` (aka `LongTensor`) type. And that's why our tests also check that, code casts to LongTensor, etc

File: tests/a_scripts/data_prep/test_convert_text_to_mds.py (hunk 11 above)
Comment (snarayan21) [reply tests/a_scripts/data_prep/test_convert_text_to_mds.py:196]: Oh i deleted the line and didn't delete the TODO lol

File: scripts/data_prep/convert_dataset_json.py (hunk 10 above)
Comment (snarayan21) [reply scripts/data_prep/convert_dataset_json.py:168]: Oh yeah weird seems it's not even used in this script. But it's used in the HF conversion script -- the dataloader will automatically emit batches of Tensor type due to the default collate function, so we convert them back to numpy arrays here.

Comment (snarayan21) [reply scripts/data_prep/convert_dataset_json.py:168]: I thought it was a bit odd that we're using a dataloader to wrap the `ConcatTokensDataset` that's returned from `build_hf_dataset` as well. Is it mainly to get multiple workers going at once, which makes the script faster or something?

File: llmfoundry/data/text_data.py (hunk 8 above)
Comment (snarayan21) [reply llmfoundry/data/text_data.py:47]: done

File: llmfoundry/data/finetuning/tasks.py (hunk 7 above)
Comment (snarayan21) [reply llmfoundry/data/finetuning/tasks.py:507]: done

File: scripts/data_prep/convert_dataset_json.py (hunk 10 above)
Comment (snarayan21) [reply scripts/data_prep/convert_dataset_json.py:168]: deleted the `generate_samples` function from `convert_dataset_json.py`

File: llmfoundry/data/text_data.py (hunk 9 above)
Comment (dakinggg) [reply llmfoundry/data/text_data.py:219]: ah got it

File: scripts/data_prep/convert_dataset_json.py (hunk 10 above)
Comment (dakinggg) [reply scripts/data_prep/convert_dataset_json.py]: yeah i think it was for leveraging torch for multiprocessing and what not. but these scripts were written a long time ago and probably have room for improvement :)

**Comment by snarayan21** [comment]
Ran the following scripts successfully: `convert_dataset_hf.py`, `convert_dataset_json.py`, `convert_finetuning_dataset.py`, `convert_text_to_mds.py`. Updated data prep readme to have instructions for `convert_text_to_mds.py`.

Using the shakespeare text file [here](https://ocw.mit.edu/ans7870/6/6.006/s08/lecturenotes/files/t8.shakespeare.txt), models trained with/without this branch have deterministic loss curves. One set of runs was with global batch size 32, the other with global batch size 256. See wandb project [here](https://wandb.ai/mosaic-ml/foundry-data-prep-runs?nw=nwusersaaketh).

Foundry regression tests are partially borked right now because of a small bug that's getting addressed in the release branch, but the tests that did run all succeeded. See [here](https://databricks.slack.com/archives/C05T1A4UMT8/p1717694690095049).

**Review by dakinggg — APPROVED** [review]
Thanks @snarayan21! And thanks for leaving the documentation better than you found it :)
